iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/twin/TwinPnPTests.java

```diff
@@ -0,0 +1,234 @@
+import com.microsoft.azure.sdk.iot.service.Module;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.*;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.IotHubTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.StandardTierHubOnlyTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.Tools;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.util.*;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.*;
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.MQTT_WS;
+import static com.microsoft.azure.sdk.iot.service.auth.AuthenticationType.SAS;
+import static com.microsoft.azure.sdk.iot.service.auth.AuthenticationType.SELF_SIGNED;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test class containing all tests to be run for PnP.
+ */
+@IotHubTest
+@StandardTierHubOnlyTest
+@RunWith(Parameterized.class)
+public class TwinPnPTests extends IntegrationTest
+{
+    protected static String iotHubConnectionString = "";
+    private static RegistryManager registryManager;
+    private String ModelId;
+
+    @Parameterized.Parameters(name = "{0}_{1}_{2}")
+    public static Collection inputs() throws IOException
+    {
+        iotHubConnectionString = Tools.retrieveEnvironmentVariableValue(TestConstants.IOT_HUB_CONNECTION_STRING_ENV_VAR_NAME);
+        isBasicTierHub = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_BASIC_TIER_HUB_ENV_VAR_NAME));
+
+        registryManager = RegistryManager.createFromConnectionString(iotHubConnectionString);
+
+        X509CertificateGenerator certificateGenerator = new X509CertificateGenerator();
+        String publicKeyCert = certificateGenerator.getPublicCertificate();
+        String privateKey = certificateGenerator.getPrivateKey();
+        String x509Thumbprint = certificateGenerator.getX509Thumbprint();
+
+        List inputs = new ArrayList();
+        inputs.addAll(Arrays.asList(
+                new Object[][]
+                        {
+                                //sas token device client, no proxy
+                                {MQTT, SAS, ClientType.DEVICE_CLIENT, publicKeyCert, privateKey, x509Thumbprint},
+                                {MQTT_WS, SAS, ClientType.DEVICE_CLIENT, publicKeyCert, privateKey, x509Thumbprint},
+
+                                //x509 device client, no proxy
+                                {MQTT, SELF_SIGNED, ClientType.DEVICE_CLIENT, publicKeyCert, privateKey, x509Thumbprint},
+
+                                //sas token device client, with proxy
+                                {MQTT_WS, SAS, ClientType.DEVICE_CLIENT, publicKeyCert, privateKey, x509Thumbprint},
+                        }
+        ));
+
+        if (!isBasicTierHub)
```

Comment:
if basic hub, will none of the tests work?

Reply:
In basic hub, modules are not supported.
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-scaling